parsers/nmparser/parse_lst_file.go

```diff
@@ -440,28 +441,30 @@ func calculateConditionNumber(lines []string, start int) float64 {
 		}
 	}
 
-	// go until another blank line and build eigenvalues vector
+	// go until non-blank line and build eigenvalues vector
 	var eigenvalues []float64
-	for i, line := range lines[start:] {
-		for _, s := range strings.Fields(line) {
-			eigenvalue, err := strconv.ParseFloat(s, 64)
-			if err == nil {
-				eigenvalues = append(eigenvalues, eigenvalue)
-			}
-		}
-
+	eigenstarted := false
+	for _, line := range lines[start:] {
 		sub := strings.TrimSpace(line)
 		if len(sub) == 0 {
-			start = start + i + 1
-			break
+			if eigenstarted {
+				break
+			} else {
+				continue
+			}
 		}
-	}
 
-	ratio := 1.0 // If only 1 eigenvalue, the Condition Number is 1.0
-	if len(eigenvalues) >= 2 {
-		sort.Float64s(eigenvalues)
-		ratio = eigenvalues[len(eigenvalues)-1] / eigenvalues[0]
+		eigenstarted = true
+		for _, s := range strings.Fields(line) {
+			eigenvalue, err := strconv.ParseFloat(s, 64)
+			if err != nil {
+				panic(fmt.Sprintf("Attempting to calculate condition number but could not parse eigenvalues -- %v", err))
```

Comment:
generally, we never want to panic unless there is a really bad reason to. Unfortunately, throughout this codebase, and some of pkgr, this was not always well held as panic is an easy escape hatch of just 💣 .

If a function is going to panic on such a situation, it should communicate it in the function name, generally using the term `must` , so like `mustParse`

If not, instead can have the function signature return the value + error, so then the if err != nil just return right there like 0, err

The idea becomes then that error can bubble all the way up the stack to the final output, or be intercepted and handled otherwise, but that is something for some other place in the code to decide what to do about it.

Reply:
I actually started to code this up and then realized that nothing for about three levels above this expects an error to come through so I just got lazy and made it panic. My thinking was that this should never be a user-facing error. In other words, it should only trigger if we parse something incorrectly, in which case spewing the ugly stack trace is probably as good as anything because we'll just want them to come to us with it.

Alternatively, we could handle it a few levels up and just log the error and return something like `nil` for the condition number. But that didn't feel right to me, honestly. Your thoughts?

Reply:
yeah i guessed as much, and could also see just using the mustParse() nomenclature and leaving the panic. I would also pass the specific value that failed on parsing as well (s) to help with diagnostics.